File: 1680. concatenation-of-consecutive-binary-numbers.java
class Solution {
    public int concatenatedBinary(int n) {
        final int MOD = 1_000_000_007;
        long res = 0;
        int bits = 0;

        for(int i=1; i<=n; i++){
            if((i & (i-1)) == 0) bits++;
            res = ((res<<bits) | i) % MOD;
        }
        return (int)res;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1:
        int n1 = 1;
        int output1 = sol.concatenatedBinary(n1);
        System.out.println("Example 1: Input: n = " + n1);
        System.out.println("Output: " + output1);
        System.out.println("Expected: 1");
        System.out.println();
        
        // Example 2:
        int n2 = 3;
        int output2 = sol.concatenatedBinary(n2);
        System.out.println("Example 2: Input: n = " + n2);
        System.out.println("Output: " + output2);
        System.out.println("Expected: 27");
        System.out.println();
        
        // Example 3:
        int n3 = 12;
        int output3 = sol.concatenatedBinary(n3);
        System.out.println("Example 3: Input: n = " + n3);
        System.out.println("Output: " + output3);
        System.out.println("Expected: 505379714");
    }
}

File: 1680. concatenation-of-consecutive-binary-numbers.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        MOD = 10**9 + 7
        res = 0
        bits = 0

        for i in range(1, n+1):
            if (i & (i-1)) == 0:
                bits += 1
            res = ((res << bits) | i) % MOD

        return res

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1:
    n = 1
    output = sol.concatenatedBinary(n)
    print(f"Example 1: Input: n = {n}")
    print(f"Output: {output}")
    print(f"Expected: 1")
    print()
    
    # Example 2:
    n = 3
    output = sol.concatenatedBinary(n)
    print(f"Example 2: Input: n = {n}")
    print(f"Output: {output}")
    print(f"Expected: 27")
    print()
    
    # Example 3:
    n = 12
    output = sol.concatenatedBinary(n)
    print(f"Example 3: Input: n = {n}")
    print(f"Output: {output}")
    print(f"Expected: 505379714")